File: LeetCode/BinaryTreeBFS/python/binaryTreeBFS-II.py
# Definition for a Node.
from collections import deque
from typing import Dict, List, Optional


class Node:
    def __init__(self, val: int = 0, left: 'Node' = None, right: 'Node' = None, next: 'Node' = None):
        self.val = val
        self.left = left
        self.right = right
        self.next = next

class Solution:
    def connect(self, root: 'Node') -> 'Node':
        if root is None:
            return None
        
        # Using BFS iterative
        queue = deque([root])

        while len(queue) > 0:
            pre = None
            qLen = len(queue)
            while qLen > 0:
                cur = queue.popleft()

                if pre is not None:
                    pre.next = cur

                if cur is not None and cur.left is not None:
                    queue.append(cur.left)

                if cur is not None and cur.right is not None:
                    queue.append(cur.right)

                pre = cur

                qLen -= 1

        return root

    @staticmethod
    def testSolution(input: Dict[str, List[str | int]]) -> None:
        # Create the binary tree using BFS
        rBinTree = Solution.BFSCreateBinaryTree(input["root"])
        expectedBinTree = Solution.BFSCreateBinaryTree(input["expected"])

        # Print the binary trees
        print("Input: r: " + Solution.BFSBinaryTreeToStr(rBinTree))
        print("Expected: " + Solution.BFSBinaryTreeToStr(expectedBinTree))

        # Call the function to be tested
        res = Solution().connect(rBinTree)

        # Print the result
        print("Result: " + Solution.BFSBinaryTreeOutputToStr(res))

        # Validate the result
        print("PASS" if Solution.ValidateResult(res, expectedBinTree) else "FAIL")
        print("-" * 50)

    @staticmethod
    def BFSCreateBinaryTree(listVal: List[str | int]) -> Node:
        arLen = len(listVal)

        if arLen < 1:
            return None
        
        root =  Node(listVal[0], None, None)
        i = 1

        # Using queue (BFS) to build the tree level by level
        queue = deque([root])

        while len(queue) > 0 and i < arLen:
            curNode = queue.popleft()

            if curNode is not None:
                # left child
                if i < arLen and listVal[i] != "null":
                    curNode.left = Node(listVal[i], None, None)
                    queue.append(curNode.left)
                i += 1

                # right child
                if i < arLen and listVal[i] != "null":
                    curNode.right = Node(listVal[i], None, None)
                    queue.append(curNode.right)
                i += 1

        return root
    
    @staticmethod
    def BFSBinaryTreeToStr(root: Optional[Node]) -> str:

        if root is None:
            return "[]"
        
        sb = []

        # Using queue (BFS) to build the tree level by level
        queue = deque([root])


        while len(queue) > 0:
            curNode = queue.popleft()

            tmp = str(curNode.val) if curNode is not None else "null"

            sb.append(tmp)
 
            if curNode is not None:
                queue.append(curNode.left)
                queue.append(curNode.right)

        new_len = 0

        # Check starting from end of sb_array where the value is not "null"
        for i in range(len(sb) - 1, -1, -1):
            if sb[i] == "null":
                continue
            else:
                new_len = i
                break

        # Remove those extra "null"
        if len(sb) > 0:
            sb = sb[0 : new_len + 1]

        return "[" + ", ".join(sb) + "]"    
    
    @staticmethod
    def BFSBinaryTreeOutputToStr(root: Optional[Node]) -> str:

        if root is None:
            return "[]"
        
        sb = []

        # Using queue (BFS) to build the tree level by level
        queue = deque([root])


        while len(queue) > 0:
            curNode = queue.popleft()

            curNodeLine = curNode

            while curNodeLine is not None:
                # append the val
                tmp = str(curNodeLine.val)
                sb.append(tmp)

                if curNodeLine.next is None:
                    # Add "#" to signify end of each level
                    sb.append("#")
                    break
                else:
                    curNodeLine = curNodeLine.next
 
            if curNode is not None:
                queue.append(curNode.left)

        return "[" + ", ".join(sb) + "]"
    
    @staticmethod
    def ValidateResult(res: Optional[Node], expected: Optional[Node]) -> bool:

        resStr = Solution.BFSBinaryTreeOutputToStr(res)
        expectedStr = Solution.BFSBinaryTreeToStr(expected)

        return resStr == expectedStr

if __name__ == "__main__":
    input = {"root": [1,2,3,4,5,"null",7], "expected": [1,"#",2,3,"#",4,5,7,"#"]}
    Solution.testSolution(input)

    input = {"root": [], "expected": []}
    Solution.testSolution(input)    

File: LeetCode/BinaryTreeBFS/JavaScript/binaryTreeBFS-II.js
// Given a binary tree

// struct Node {
//   int val;
//   Node *left;
//   Node *right;
//   Node *next;
// }
// Populate each next pointer to point to its next right node. If there is no next right node, the next pointer should be set to NULL.

// Initially, all next pointers are set to NULL.

// Example 1:

// Input: root = [1,2,3,4,5,null,7]
// Output: [1,#,2,3,#,4,5,7,#]
// Explanation: Given the above binary tree (Figure A), your function should populate each next pointer to point to its next right node, just like in Figure B. The serialized output is in level order as connected by the next pointers, with '#' signifying the end of each level.
// Example 2:

// Input: root = []
// Output: []

// Constraints:

// The number of nodes in the tree is in the range [0, 6000].
// -100 <= Node.val <= 100

// Follow-up:

// You may only use constant extra space.
// The recursive approach is fine. You may assume implicit stack space does not count as extra space for this problem.

/**
 * // Definition for a _Node.
 * function _Node(val, left, right, next) {
 *    this.val = val === undefined ? null : val;
 *    this.left = left === undefined ? null : left;
 *    this.right = right === undefined ? null : right;
 *    this.next = next === undefined ? null : next;
 * };
 */

const Result = { PASS: "\x1b[92mPASS\x1b[0m", FAIL: "\x1b[91mFAIL\x1b[0m" };

class _Node {
  constructor(val, left, right, next) {
    this.val = val === undefined ? null : val;
    this.left = left === undefined ? null : left;
    this.right = right === undefined ? null : right;
    this.next = next === undefined ? null : next;
  }
}

class ConnectRecord {
  constructor(values, expected) {
    this.values = values;
    this.expected = expected;
  }
}

class Solution {
  /**
   * @param {_Node} root
   * @return {_Node}
   */
  connect(root) {
    if (!root) return root;

    // Using BFS
    let queue = [];

    queue.push(root);

    while (queue.length !== 0) {
      let pre = null;
      let qLen = queue.length;
      while (qLen !== 0) {
        let cur = queue.shift();

        if (pre) {
          pre.next = cur;
        }

        if (cur && cur.left) {
          queue.push(cur.left);
        }
        if (cur && cur.right) {
          queue.push(cur.right);
        }

        pre = cur;
        // Don't forget to decrement qLen!
        qLen -= 1;
      }
    }

    return root;
  }
}

function createTree(values) {
  if (!values.length || values[0] === null) return null;

  const root = new _Node(values[0]);
  const queue = [root];
  let index = 1;

  while (queue.length && index < values.length) {
    const node = queue.shift();

    if (values[index] !== null && values[index] !== undefined) {
      node.left = new _Node(values[index]);
      queue.push(node.left);
    }
    index++;

    if (index < values.length && values[index] !== null && values[index] !== undefined) {
      node.right = new _Node(values[index]);
      queue.push(node.right);
    }
    index++;
  }

  return root;
}

// Serialize in level order as connected by the next pointers,
// with '#' signifying the end of each level.
function serializeByNext(root) {
  const output = [];
  let levelStart = root;

  while (levelStart) {
    let node = levelStart;
    let nextStart = null;

    while (node) {
      output.push(node.val);
      if (!nextStart) nextStart = node.left || node.right;
      node = node.next;
    }

    output.push("#");
    levelStart = nextStart;
  }

  return output;
}

function testSolution(record) {
  const solution = new Solution();
  const result = serializeByNext(solution.connect(createTree(record.values)));
  const pass = JSON.stringify(result) === JSON.stringify(record.expected);

  console.log(`Input: root = ${JSON.stringify(record.values)}`);
  console.log(`Expected: [${record.expected.join(",")}]`);
  console.log(`Result: [${result.join(",")}]`);
  console.log(pass ? Result.PASS : Result.FAIL);
}

const records = [
  new ConnectRecord([1, 2, 3, 4, 5, null, 7], [1, "#", 2, 3, "#", 4, 5, 7, "#"]),
  new ConnectRecord([], []),
  new ConnectRecord([1], [1, "#"]),
  new ConnectRecord([1, 2, 3], [1, "#", 2, 3, "#"]),
];

records.forEach((record, index) => {
  console.log(`# Test case ${index + 1}`);
  testSolution(record);
  console.log("----------------------------------------");
});
